### ISSUE TITLE
data_loading_tutorial.py iterators

### ISSUE THREAD

**Issue by patricknaughton01** [pytorch/tutorials#995]
I like this tutorial but I think it would be better if it included an example of how to define __next__() and __iter__() methods so that the dataset can be used with `enumerate`.

**Comment by noqqaqq** [comment]
Some thoughts looking at this issue:
1. __iter__ and __next__ are not needed to use enumerate - it works out of the box with __len__ and __getitem__. Tutorial can be then easily improved to use enumerate in relevant places without reimplementing dataset class.
2. There is [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset). I guess, implementing it would be most sufficient for this example.
3. Using IterableDataset raise some issues:
- random access is lost, which then requires looping or skipping iterator for [this](https://github.com/pytorch/tutorials/blob/dd6a55da69488e104c56061aad627987628287cb/beginner_source/data_loading_tutorial.py#L322) case (need to make sure to implement it in a way that minimize perf penalty)
- IterableDataset does not allow for shuffling, which is needed when [passing dataset to Dataloader](https://github.com/pytorch/tutorials/blob/dd6a55da69488e104c56061aad627987628287cb/beginner_source/data_loading_tutorial.py#L383). There are ways to implement this like described [here](https://discuss.pytorch.org/t/how-to-shuffle-an-iterable-dataset/64130/6) or [here](https://github.com/pytorch/pytorch/issues/62310) or [here](https://discuss.pytorch.org/t/how-to-shuffle-multi-worker-iterabledataset-loading/142203/2), but I don't know yet how to do it in the simplest correct way for the sake of tutorial.

To sum up, I'll follow with pushing patch for 1. and try to copy and modify dataset tutorial to cover iterable case (having both in single file in my opinion would obfuscate the examples and make them hard to track / understand).

If anyone has comments on this or hints on shuffling please let me know.

**Comment by NicolasHug** [comment]
I agree with @noqqaqq's thoughts that there's no need for `next()` and `iter()` to be present for `enunerate()` to work. The original point of the issue may have been to include an example for **iterable** datasets whoc have a `__next__()` method instead of `__getitem__()`, but like @noqqaqq also correctly pointed out, there are other resources out there to do that and that's not the original point of this tutorial. 

For the simple purpose of exposing how to use `enumerate()`, https://github.com/pytorch/tutorials/pull/2407 LGTM
